opm/autodiff/WellStateFullyImplicitBlackoil.hpp

```diff
@@ -793,6 +799,49 @@ namespace Opm
         std::vector<int> top_segment_index_;
         int nseg_; // total number of the segments
 
+        ::Opm::data::Segment
+        reportSegmentResults(const PhaseUsage& pu,
+                             const int         wellID,
+                             const int         segmentID) const
+        {
+            auto segRes = ::Opm::data::Segment{};
+
+            const auto segDoF =
+                this->topSegmentIndex(wellID) + segmentID;
+
+            const auto* rate =
+                &this->segRates()[segDoF * this->numPhases()];
+
+            segRes.pressure = this->segPress()[segDoF];
+
+            if (pu.phase_used[Water]) {
+                segRes.rates.set(data::Rates::opt::wat,
+                                 rate[pu.phase_pos[Water]]);
+            }
+
+            if (pu.phase_used[Oil]) {
+                segRes.rates.set(data::Rates::opt::oil,
+                                 rate[pu.phase_pos[Oil]]);
+            }
+
+            if (pu.phase_used[Gas]) {
+                segRes.rates.set(data::Rates::opt::gas,
+                                 rate[pu.phase_pos[Gas]]);
+            }
+
+            segRes.segNumber = segmentID + 1;
```

Comment:
I am not saying this is wrong, because it really depends on how you use it. 

More like a reminder,  from the input side, `segment number` is more or less a ID number, which is not necessarily related to its storage index.  

But it really depends how to define and use it. Please refer to the function `segmentNumberToIndex()` in `WellSegments` for reference.

Reply:
> capture the reverse mapping (index -> segment number)

With index, you can find the `Segment`, `Segment` has the information of `segment number`.

Reply:
>  I'll either have to capture the reverse mapping (index -> segment number) as a data member of WellStateFullyImplicitBlackoil or change the interface of report() to somehow enable passing the WellSegments from the outside.

I think this is right.

Reply:
> > 'll either have to capture the reverse mapping (index -> segment number) as a data member of `WellStateFullyImplicitBlackoil` or change the interface of `report()` to somehow enable passing the `WellSegments` from the outside.
> 
> I think this is right.

Okay, in that case I think I'll add the data member because it can be built as part of `initWellStateMSWell()` and therefore serve as a consistency check for whether or not that function has been called.

Reply:
> Okay, in that case I think I'll add the data member because it can be built as part of `initWellStateMSWell()` 

Sure. 

> therefore serve as a consistency check for whether or not that function has been called.

I am not sure what this means, but I do not think it matters much here.

Reply:
> > therefore serve as a consistency check for whether or not [`initWellStateMSWell()`] has been called.
> 
> I am not sure what this means, but I do not think it matters much here.

It'll be a signal that we should not attempt to return segment-related vectors for anything other than the top segment unless the user specifies [`--use-multisegment-well`](https://github.com/OPM/opm-simulators/blob/33da0aea09158408563a84393400f08efd41c38a/opm/autodiff/BlackoilModelParametersEbos.hpp#L68).  Otherwise we'd be indexing out of bounds since the `nseg_` member is uninitialised in that case and my new [`numSegments(wellID)`](https://github.com/OPM/opm-simulators/blob/28cb22d5d08fab9ebfb9d5ad9613241cc1adc9b7/opm/autodiff/WellStateFullyImplicitBlackoil.hpp#L837-L844) function would return useless values.

Reply:
> Otherwise we'd be indexing out of bounds since the nseg_ member is uninitialised in that case and my new numSegments(wellID) function would return useless values.

You can do it in your way. If you think it is a good idea to add a trivial initialization of `nseg_` similar to `top_segment_index_` at the end of  function `void init()`.  Leaving `nseg_` initialized was a mistake.